packages/gatsby/src/redux/run-sift.js

```diff
@@ -83,89 +89,287 @@ function handleMany(siftArgs, nodes, sort, resolvedFields) {
+ *
+ * @param args raw graphql query filter/sort as an object
+ * @property {boolean | number | string} args.type gqlType. See build-node-types
+ * @property {boolean} args.firstOnly true if you want to return only the first
+ *   result found. This will return a collection of size 1. Not a single element
+ * @property {{filter?: Object, sort?: Object} | undefined} args.queryArgs
+ * @property {undefined | Map<string, Map<string | number | boolean, Node>>} args.typedKeyValueIndexes
+ *   May be undefined. A cache of indexes where you can look up Nodes grouped
+ *   by a key: `types.join(',')+'/'+filterPath.join('+')`, which yields a Map
+ *   which holds a Set of Nodes for the value that the filter is trying to eq
+ *   against. If the property is `id` then there is no Set, it's just the Node.
+ *   This object lives in query/query-runner.js and is passed down runQuery
+ * @returns Collection of results. Collection will be limited to 1
  *   if `firstOnly` is true
  */
-const runSift = (args: Object) => {
-  const { getNode, addResolvedNodes, getResolvedNode } = require(`./nodes`)
-
-  const { nodeTypeNames } = args
-  if (
-    args.queryArgs?.filter &&
-    Object.getOwnPropertyNames(args.queryArgs.filter).length === 1 &&
-    typeof args.queryArgs.filter?.id?.eq === `string`
-  ) {
-    // The args have an id.eq which subsumes all other queries
-    // Since the id of every node is unique there can only ever be one node found this way. Find it and return it.
-    let id = args.queryArgs.filter.id.eq
-    let node = undefined
-    nodeTypeNames.some(typeName => {
-      node = getResolvedNode(typeName, id)
-      return !!node
-    })
-    if (node) {
-      return [node]
+const runFilterAndSort = (args: Object) => {
+  const {
+    queryArgs: { filter, sort } = { filter: {}, sort: {} },
+    resolvedFields = {},
+    firstOnly = false,
+    nodeTypeNames,
+    typedKeyValueIndexes,
+  } = args
+
+  let result = applyFilters(
+    filter,
+    firstOnly,
+    nodeTypeNames,
+    typedKeyValueIndexes,
+    resolvedFields
+  )
+
+  return sortNodes(result, sort, resolvedFields)
+}
+
+exports.runSift = runFilterAndSort
+
+/**
+ * Applies filter. First through a simple approach, which is much faster than
+ * running sift, but not as versatile and correct. If no nodes were found then
+ * it falls back to filtering through sift.
+ *
+ * @param {Object | undefined} filter
+ * @param {boolean} firstOnly
+ * @param {Array<string>} nodeTypeNames
+ * @param {undefined | Map<string, Map<string | number | boolean, Node>>} typedKeyValueIndexes
+ * @param resolvedFields
+ * @returns {Array<Node> | undefined} Collection of results. Collection will be
+ *   limited to 1 if `firstOnly` is true
+ */
+const applyFilters = (
+  filter,
+  firstOnly,
+  nodeTypeNames,
+  typedKeyValueIndexes,
+  resolvedFields
+) => {
+  let result
+  if (typedKeyValueIndexes) {
+    result = filterWithoutSift(filter, nodeTypeNames, typedKeyValueIndexes)
+    if (result) {
```

Comment:
Wouldn't that mean that we always repeat search if something isn't found?

Reply:
Yes. There are some cases where this approach can miss, like (apparently) using `@proxy` to rename a property. There's a test that failed because of this and that's how I know.
This means there's a little overhead for these cases. We can improve on many of these cases, I'm sure. for example we can analyze the filter at this point and determine whether or not the miss is expected